### ISSUE TITLE
RestrictPtrTraits in CUDA potentially has no effect.

### ISSUE THREAD

**Issue by lkskstlr** [pytorch/pytorch#76632]
### 🐛 Describe the bug

Potentially `torch::PackedTensorAccessor32<T, N, torch::RestrictPtrTraits>` might be treated the same as `torch::PackedTensorAccessor32<T, N>` in CUDA given how `__restirct__` is handeled.

The pointer to the data is a class member [TensorAccessor.h#L62](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/TensorAccessor.h#L62)
```cpp
// The PtrTraits argument is only relevant to cuda to support `__restrict__` pointers.
template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits, typename index_t = int64_t>
class TensorAccessorBase {
public:
 // code omitted for readability
protected:
  PtrType data_;
  const index_t* sizes_;
  const index_t* strides_;
};
```

However, this is not respected by `nvcc`: [As explained by Nvidia here](https://forums.developer.nvidia.com/t/restrict-seems-to-be-ignored-for-base-pointers-in-structs-having-base-pointers-with-restrict-as-kernel-arguments-directly-works-as-expected/154020/13?u=lkskstlr).

I made an example to show this in the [compiler explorer](https://godbolt.org/z/9c3z8GfT8).

I tried to make an example that shows an actual performance difference but have been struggling so far. It seems that caching can alleviate the performance problem in easy cases.

Thank you all for the amazing library and sorry if this is not a bug but a misunderstanding on my side :slightly_smiling_face: 

### Versions

Does not apply.

**Comment by lkskstlr** [comment]
Hi @ngimel,

I would be happy about your opinion on this, but for me it's totally fine to just close this. I am a big PyTorch fan and it seems that there are enough open issues already.

Best
Lukas

**Comment by ngimel** [comment]
Yeah, it's a legit inefficiency, but as you say, it doesn't show up as a perf regression in most cases, and doesn't lead to incorrect results, so unless you really want to fix it (we'll accept a PR) it's unlikely to be fixed.

**Comment by f4ded4k** [comment]
Hi @lkskstlr 
The simplest fix seems to be to decouple the data pointer from the indexing information, have accessor API return both (a_ptr, a_accessor), pass pointers with __restrict__ along with accessor struct to the kernel, define free functions for indexing i.e. get(a_ptr, i, j, k) instead of a[i][j][k].
